samples/42.scaleout/src/main/java/com/microsoft/bot/sample/scaleout/MemoryStore.java

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.microsoft.bot.sample.scaleout;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.microsoft.bot.schema.Pair;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Semaphore;
+
+/**
+ * A thread safe implementation of the IStore abstraction intended for testing.
+ */
+public class MemoryStore implements Store {
+
+    private Map<String, Pair<JsonNode, String>> store = new HashMap<>();
+
+    // When setting up the database, calls are made to CosmosDB. If multiple calls are made, we'll end up setting the
+    // collectionLink member variable more than once. The semaphore is for making sure the initialization of the
+    // database is done only once.
+    private static final Semaphore semaphore = new Semaphore(1);
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public CompletableFuture<Pair<JsonNode, String>> load(String key) {
+        try {
+            semaphore.wait();
+
+            Pair<JsonNode, String> value = store.get(key);
+            if (value != null) {
+                return CompletableFuture.completedFuture(value);
+            }
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        } finally {
+            semaphore.release();
+            return CompletableFuture.completedFuture(new Pair<>(null, null));
```

Comment:
I'm not quite sure about this `return` in the `finally` block, doesn't it always run?

Reply:
That's what I want, it's the same behavior as C#, otherwise it will return depending on the `if/else` statements

Reply:
But I think that in this case always will return `return CompletableFuture.completedFuture(new Pair<>(null, null));` despite the return in the `try` block, see [this](https://beginnersbook.com/2013/05/java-finally-return/). In C# both returns are in the `try` block.

Reply:
Mm no, depending on the `if/else` statement which contains another `return`, just like C#